### PR TITLE
[Core][Spec Decode] Dynamic verification for dflash

### PR DESCRIPTION
**Pull request by lrioxh**
## Purpose

DFlash generates multiple speculative tokens in parallel within one step. This can substantially increase the speculative length, leading to longer accepted lengths and better performance. However, a longer verification length also puts computing pressure on the target model. Compared with methods such as MTP, DFlash have a lower acceptance rate, meaning many verified speculative tokens are eventually discarded.

This PR adds a simple dynamic verification mechanism for DFlash. Instead of always verifying the full preconfigured speculative length, we can truncate speculative tokens early based on draft-token confidence. The motivation is that low-confidence draft tokens are unlikely to be accepted, so removing them before target-model verification can reduce wasted compute.

This is especially useful under high-concurrency workloads, where target-model verification is more likely to become compute-bound. In low-concurrency settings, the extra computation, copying and shorter accepted length can hurt performance. For details, please see the test result.

Main changes:

- Add `dynamic_verifying` to `SpeculativeConfig`.
  - `None`: disabled.
  - `float` in `(0, 1)`: confidence threshold.
  - `"auto"`: adaptive per-request `threshold = clamp(mean(conf), 0.1, 0.8)`.
- Add `dynamic_verifying_min_length` to avoid over-truncating every request.
- Compute DFlash draft-token confidence from greedy draft logits.
- Truncate DFlash speculative tokens once confidence drops below the configured threshold.
- Propagate the valid draft-token counts back to the scheduler so speculative decoding stats use the actually verified draft length.
- Document the new DFlash speculative config options.

AI assistance was used in preparing this PR. I reviewed the changed lines and am responsible for the implementation, behavior, and tests.

## Test Plan

GPU: H800

Functional coverage:

- Dynamic verification disabled.
- Dynamic verification enabled with varied thresholds and “auto”.
- Different `dynamic_verifying_min_length` values.
- Async scheduling disabled and enabled.
- Tensor parallelism coverage.
- Model coverage:
  - Qwen3-8B + DFlash.
  - Qwen3.5-9B + DFlash.

Performance coverage:

- Qwen3-8B on GSM8K.
```bash
vllm serve ...\
    --gpu-memory-utilization 0.9 \
    --tensor-parallel-size ${TENSOR_PARALLEL_SIZE:=1} \
    --compilation-config '{"cudagraph_mode":"FULL"}' \
    --attention-backend flash_attn \
    --no-enable-prefix-caching \
    --speculative-config '{"method": "dflash", "model": "...", "num_speculative_tokens": 15, "dynamic_verifying": $method, "dynamic_verifying_min_length": 1}' 

```
- Qwen3.5-9B on longbench.

```bash
vllm serve ...\
    --gpu-memory-utilization 0.9 \
    --tensor-parallel-size ${TENSOR_PARALLEL_SIZE:=2} \
    --compilation-config '{"cudagraph_mode":"PIECEWISE"}' \
    --attention-backend flash_attn \
    --no-enable-prefix-caching \
    --speculative-config '{"method": "dflash", "model": "...", "num_speculative_tokens": 15, "dynamic_verifying": $method, "num_speculative_tokens_per_batch_size": [[1, 32, 15], [33, 96, 7], [97, 128, 3]]}' 

```

## Test Result
Functional completeness testing did not find correctness or stability issues.
#### Qwen3-8B + DFlash on GSM8K

The accepted length and verifying length are stable across different batch sizes under the same confidence threshold. Therefore, the table below reports the mean values across all tested batch sizes (`bs=1,8,16,32,64,128`) to summarize the effect of each threshold. We set the max input tokens to 2048, with an average output length of approximately 300.

| Conf thr | Mean acc length | Mean verifying length |
| --- | ---: | ---: |
| 0 (base) | 4.81 | 16.00 |
| 0.1 | 4.81 | 14.82 |
| 0.3 | 4.73 | 9.79 |
| 0.5 | 4.49 | 7.01 |
| 0.7 | 3.99 | 5.10 |
| auto | 4.25 | 5.88 |

image

#### Qwen3.5-359B on longbench

image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

</details>

### CODE REVIEW

File: vllm/v1/spec_decode/dflash.py [hunk 1]
```diff
@@ -298,3 +304,70 @@ def _get_eagle3_use_aux_hidden_state_from_config(self):
...
+    def _greedy_sample(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        if self.dynamic_verifying_method:
+            logits = self.model.compute_logits(hidden_states)
+            max_logits, draft_token_ids = logits.max(dim=-1)
+            # Numerically stable max-softmax: exp(max - logsumexp) equals
+            # the max probability from softmax without materializing the
+            # full softmax tensor.
+            log_sum_exp = logits.logsumexp(dim=-1)
+            draft_confidence = (max_logits - log_sum_exp).exp()
+            draft_confidence = draft_confidence.view(-1, self.num_speculative_tokens)
```
[the remark this answers is not in the document]
Comment (lrioxh) [reply vllm/v1/spec_decode/dflash.py]: number of draft tokens is equal to num_speculative_tokens

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -829,15 +829,17 @@ def __init__(
...
+            self.speculative_config.use_ngram_gpu()
+            or self.speculative_config.dynamic_verifying
         ):
             self._num_valid_draft_tokens_cpu = torch.empty(
                 self.max_num_reqs, dtype=torch.int32, pin_memory=self.pin_memory
             )
             self._num_valid_draft_tokens_event = torch.cuda.Event()
             self._num_valid_draft_tokens_copy_stream = torch.cuda.Stream()
+            if self.speculative_config.dynamic_verifying:
+                self._dynamic_truncated_spec_tokens: dict[str, int] = {}
```
[the remark this answers is not in the document]
Comment (lrioxh) [reply vllm/v1/worker/gpu_model_runner.py]: corrected

**Comment by benchislett** [comment]
I think this is a duplicate of this PR: https://github.com/vllm-project/vllm/pull/35301
For which I have some reservations. I expect the same concerns will apply to this PR, see my comments.

**Comment by lrioxh** [comment]
> I think this is a duplicate of this PR: #35301 For which I have some reservations. I expect the same concerns will apply to this PR, see my comments.

Thanks for your feedback!

I took a look at #35301. I agree that both PRs share a similar idea around confidence-based truncation, but I think the goals and implementations are quite different:

- #35301 implements per-step dynamic `k` for autoregressive speculative decoding by using more spec steps with early stopping. This does not directly apply to DFlash parallel drafting.
- Simply 0-padding low-confidence tokens does not reduce verification compute of target model.
- The main goal of this PR is per-request dynamic verification length, so that the target model verifies fewer draft tokens for batched requests whose draft tokens are unlikely to be accepted.

That said, I do think these two PRs could potentially be unified at the config, runner, and scheduler layers, while still using method-specific dynamic strategies for different spec decoding methods.

Following your comments in #35301, I will add more comparisons, including:

- Qwen3-8B without speculative decoding
- DFlash with `k=7`
- EAGLE3 with `k=2 or 3`
- MATH500 results

The current results also do show that this approach does not provide speedup at lower batch sizes.

I also looked at #32374. That PR goes deeper into dynamically allocating `k` in scheduler, which is also something I would like to explore next. More specifically, I am considering a scheduler policy based on `batched_tokens * context_len`, combined with confidence-based truncation. Similar to #35301, I think #32374 could also be complementary to this PR.

Regarding the CUDA graph concern, I guess it may depend on specific operators? In the Qwen3-8B experiments, shorter dynamic verification length does improve performance, but this may not hold for other operators such as linear attn or moe. I will test more situations to better understand this. In addition, the case of high concurrency, for example 64, batched_token = 64*16 = 1024 is already greater than the general max_graph_size = 512, and cudagraph_mode becomes None. However, I am not sure whether there are other optimizations for uniform_decode besides CUDA graph. If there are, please point them out and I will further study them.

**Comment by benchislett** [comment]
Thanks for the summary! A few things to watch out for:
- Some attention backends may rely on having a fixed number of spec tokens in the batch to dispatch an optimized kernel when the requests are all verification requests. If we vary the number of verification tokens per step, we need to ensure that we're still issuing the optimized kernel pathway. This is largely optimized for server-class Blackwell GPUs and may not be visible on others. https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/flashinfer.py#L662C14-L662C43
- - This is a particular challenge for FULL CUDA graph compatibility: I believe we currently record graphs across batch sizes with a uniform verification size equal to `1 + num_spec_tokens`. If we replay the wrong graph we might hit a wrong attention kernel and corrupt the outputs.
- Synchronization overhead is nontrivial. We have worked hard to ensure that there are no blocking synchronization points in our main model runner loop. By deciding the verification size based on some heuristic derived from previous-step acceptance rates, we introduce a data dependence between GPU data and the CPU decision-making, forcing a synchronization between the steps. This is costly as the GPU must go idle between the synchronization point and the next issued work, and if the CPU cannot issue work fast enough then we might waste a lot of time. This is especially critical in the fastest workloads, e.g. NVFP4 on B200/B300.

**Comment by lrioxh** [comment]
> Thanks for the summary! A few things to watch out for:
> 
> * Some attention backends may rely on having a fixed number of spec tokens in the batch to dispatch an optimized kernel when the requests are all verification requests. If we vary the number of verification tokens per step, we need to ensure that we're still issuing the optimized kernel pathway. This is largely optimized for server-class Blackwell GPUs and may not be visible on others. https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/flashinfer.py#L662C14-L662C43
> * * This is a particular challenge for FULL CUDA graph compatibility: I believe we currently record graphs across batch sizes with a uniform verification size equal to `1 + num_spec_tokens`. If we replay the wrong graph we might hit a wrong attention kernel and corrupt the outputs.
> * Synchronization overhead is nontrivial. We have worked hard to ensure that there are no blocking synchronization points in our main model runner loop. By deciding the verification size based on some heuristic derived from previous-step acceptance rates, we introduce a data dependence between GPU data and the CPU decision-making, forcing a synchronization between the steps. This is costly as the GPU must go idle between the synchronization point and the next issued work, and if the CPU cannot issue work fast enough then we might waste a lot of time. This is especially critical in the fastest workloads, e.g. NVFP4 on B200/B300.

Thanks for pointing these out. These are all important concerns.

+ Unfortunately, I do not have access to Blackwell GPUs, so I cannot directly validate the server-class Blackwell path. From the code, it looks like TRTLLM decode requires uniform query lengths per request. In that case, this dynamic verification feature may need to be disabled for that backend/path.

+ For the Qwen3-8B setup I tested most, I found that even without dynamic verification length, where the query length is still fixed at `1 + num_spec_tokens`, `desc.uniform` is already `False`, and CUDA graph mode is `mixed prefill-decode, FULL`. In that case, whether the verification query length is uniform does not seem to change the graph path. For Qwen3.5-35B, however, I observed a different behavior: CUDA graph is `FULL` when the verification size is uniform, but falls back to `PIECEWISE` when it becomes non-uniform. This does cause a performance regression with dynamic verification. I need to spend more time understanding this path. In the worst case, models/backends that rely on the uniform verification path may simply not be a good fit for this dynamic verification feature.

+ Synchronization is another important issue. I think this is one reason why this feature gets a worse performance at low batch sizes. However, synchronization seems unavoidable here. This is similar to the `ngram_gpu` path in #29184 (MRV2 #40704): the async copy and synchronize are necessary; otherwise, as we also tested, the acceptance drops. That said, I agree the current implementation can likely be improved. A better approach may be to share the synchronization point with the existing `draft_token_id` copy synchronization, similar to MRV2, instead of using an independent synchronization stream. The current implementation reuses the `ngram_gpu` valid-length synchronization path mainly for simplicity. I will try to change it to share the existing synchronization and see whether that improves performance.

Thanks again for the detailed suggestions!

**Comment by benchislett** [comment]
I feel strongly that we must fundamentally avoid synchronization whenever possible. I do not feel comfortable supporting this feature at this time.

**Comment by lrioxh** [comment]
Although closed, test results are updated here,
image

with speculative-config:
```
{
    "method": "dflash",
    "model": "...",
    "num_speculative_tokens": 15,
    "dynamic_verifying": "auto",
    "dynamic_verifying_min_batch_size": 16,
    "num_speculative_tokens_per_batch_size": [
        [1, 32, 15],
        [33, 96, 7],
        [97, 128, 3]
    ]
}
```
enabling this with #32374 together is expected to yield better performance.
